src/shared/sol-util.c

```diff
@@ -52,6 +81,104 @@ sol_util_memdup(const void *data, size_t len)
     return ptr;
 }
 
+double
+sol_util_strtodn(const char *nptr, char **endptr, size_t len, bool use_locale)
+{
+    char *tmpbuf, *tmpbuf_endptr;
+    double value;
+
+#if defined(HAVE_LOCALE) && defined(HAVE_STRTOD_L)
+    if (!use_locale) {
+        if (!init_c_locale()) {
+            /* not great, but a best effort to convert something */
+            use_locale = false;
+            SOL_WRN("could not create locale 'C', use current locale.");
+        }
+    }
+#endif
+
+    /* NOTE: Using a copy to ensure trailing \0 and strtod() so we
+     * properly parse numbers with large precision.
+     *
+     * Since parsing it is complex (ie:
+     * http://www.netlib.org/fp/dtoa.c), we take the short path to
+     * call libc.
+     */
+    tmpbuf = strndupa(nptr, len);
+
+    errno = 0;
+#ifdef HAVE_LOCALE
+    if (!use_locale) {
+#ifdef HAVE_STRTOD_L
+        value = strtod_l(tmpbuf, &tmpbuf_endptr, c_locale);
+#else
+        /* fallback to query locale's decimal point and if it's
+         * different than '.' we replace JSON's '.' with locale's
+         * decimal point if the given number contains a '.'.
+         *
+         * We also replace any existing locale decimal_point with '.'
+         * so it will return correct endptr.
+         *
+         * Extra care since decimal point may be multi-byte.
+         */
+        struct lconv *lc = localeconv();
+        if (lc && lc->decimal_point && !streq(lc->decimal_point, ".")) {
+            if (strchr(tmpbuf, '.') || strstr(tmpbuf, lc->decimal_point)) {
+                int dplen = strlen(lc->decimal_point);
+                const char *src, *src_end = tmpbuf + len;
+                char *dst;
+                if (dplen == 1) {
+                    for (src = tmpbuf, dst = tmpbuf; src < src_end; src++, dst++) {
+                        if (*src == '.')
+                            *dst = *lc->decimal_point;
+                        else if (*src == *lc->decimal_point)
+                            *dst = '.';
+                    }
+                } else {
+                    char *transl;
+                    unsigned count = 0;
+
+                    for (src = tmpbuf; src < src_end; src++) {
```

Comment:
I did this in my first version, but it may cause us problems to set the "endptr" in the same way `strtod()` would, thus I'm doing the "costly" version that replaces them all -- better safe than sorry, and it shouldn't impact much if there is a single '.'

What I could do is to use count in the second loop and stop earlier if all was done, resulting in a memcpy() of the rest. Not sure it would matter much.